### DISCUSSION TITLE
Sharing some findings for android native ikev2 client

### DISCUSSION THREAD

**Discussion by Arunmor1510** [strongswan/strongswan#1690] [Ideas]
Since Android 11, native support for ikev2 has been added. However android's built-in interface doesn't provide much configurable options. For me the biggest pain is there's no option to configure remote identity so android just can't verify the certificate (seems it uses server address as remote identity), because my certificate contains the apex domain and some wildcard subdomains, which instead is somewhat inconvenient for ikev2.

Is android's ikev2 support so primitive and I have to either stick to strongswan's android client or apply for another certificate? Fortunately after some digging, I found that android has something called [Ikev2VpnProfile](https://developer.android.com/reference/android/net/Ikev2VpnProfile) and [VpnManager](https://developer.android.com/reference/android/net/VpnManager), which provide a way for normal user apps to create and start a ikev2 vpn session on its own. [Ikev2VpnProfile](https://developer.android.com/reference/android/net/Ikev2VpnProfile) can make use of [IkeTunnelConnectionParams](https://developer.android.com/reference/android/net/ipsec/ike/IkeTunnelConnectionParams) to configure a variety of options, including remote identity, problem solved!

But not so fast. A closer look at the documentation indicates that a `Ikev2VpnProfile` can be constructed from `IkeTunnelConnectionParams` only starting from android 13, but my phone is on 12! Well, I reached here and at least l need to validate my idea. I did some quick learning, wrote some ugly code, and finally, I'm able to see the vpn connects and ipsec policy is installed (note that android emulator with some version of system images has a built-in `su` so it's easy to switch to root). There's still a issue, the whole android system may crash if I disconnect the vpn, not sure why, but for now I don't want to dig further unless my phone receives android 13 update.
In conclusion, the latest android release has a rich support for ikev2, based on which a native ikev2 client can be made, thanks for reading!
